File: microservices/order_service/project/services/booking_service.py
import requests
from flask import current_app
from datetime import datetime, timedelta

from project.models.models import Booking, BookingDetail, db


def _fetch_seat_map(showtime_id, access_token):
    try:
        response = requests.get(
            f"{current_app.config['CINEMA_SERVICE_URL'].rstrip('/')}/showtimes/{showtime_id}/seats",
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=current_app.config.get("REQUEST_TIMEOUT", 5),
        )
    except requests.RequestException as exc:
        raise ValueError("Cinema service is unavailable") from exc

    if response.status_code == 404:
        raise ValueError("Showtime not found")

    if response.status_code != 200:
        raise ValueError("Unable to verify seats from cinema service")

    return response.json()


def _lock_seats(showtime_id, seat_codes, access_token):
    """Lock specific seats for booking"""
    try:
        response = requests.post(
            f"{current_app.config['CINEMA_SERVICE_URL'].rstrip('/')}/showtimes/{showtime_id}/seats/lock",
            json={"seat_codes": seat_codes},
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=current_app.config.get("REQUEST_TIMEOUT", 5),
        )
    except requests.RequestException as exc:
        raise ValueError("Cinema service is unavailable") from exc

    if response.status_code == 404:
        raise ValueError("Showtime not found")

    if response.status_code != 200:
        error_data = response.json() if response.content else {}
        raise ValueError(error_data.get("message", "Unable to lock seats"))

    return response.json()


def _release_seats(showtime_id, seat_codes, access_token):
    """Release seat locks"""
    try:
        response = requests.post(
            f"{current_app.config['CINEMA_SERVICE_URL'].rstrip('/')}/showtimes/{showtime_id}/seats/release",
            json={"seat_codes": seat_codes},
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=current_app.config.get("REQUEST_TIMEOUT", 5),
        )
    except requests.RequestException:
        # Ignore release errors as this is cleanup
        pass


def _get_reserved_seats(showtime_id, seat_codes):
    booked_rows = (
        db.session.query(BookingDetail.seat_code)
        .join(Booking, Booking.booking_id == BookingDetail.booking_id)
        .filter(
            Booking.showtime_id == showtime_id,
            Booking.status >= 0,
            BookingDetail.seat_code.in_(seat_codes),
        )
        .all()
    )
    return {row[0] for row in booked_rows}


def create_booking(user_id, showtime_id, seat_codes, access_token):
    normalized_seat_codes = [seat_code.strip().upper() for seat_code in seat_codes if seat_code.strip()]
    unique_seat_codes = list(dict.fromkeys(normalized_seat_codes))

    if not unique_seat_codes:
        raise ValueError("At least one seat must be selected")

    # First, lock the specific seats we want to book
    try:
        _lock_seats(showtime_id, unique_seat_codes, access_token)
    except ValueError as exc:
        raise ValueError(f"Unable to lock seats: {str(exc)}")

    try:
        # Now fetch seat map to verify seats exist and get prices
        seat_map = _fetch_seat_map(showtime_id, access_token)
        available_seats = {seat["code"]: seat for seat in seat_map.get("seats", [])}

        invalid_seats = [seat_code for seat_code in unique_seat_codes if seat_code not in available_seats]
        if invalid_seats:
            raise ValueError(f"Invalid seat codes: {', '.join(invalid_seats)}")

        # Check if seats are already booked in our database
        already_booked = _get_reserved_seats(showtime_id, unique_seat_codes)
        if already_booked:
            raise ValueError(f"Seats already booked: {', '.join(sorted(already_booked))}")

        total_amount = sum(available_seats[seat_code]["price"] for seat_code in unique_seat_codes)

        booking = Booking(
            user_id=user_id,
            showtime_id=showtime_id,
            total_amount=total_amount,
            status=1,  # Status = 1 (confirmed, no payment needed for this project)
        )

        db.session.add(booking)
        db.session.flush()

        for seat_code in unique_seat_codes:
            db.session.add(
                BookingDetail(
                    booking_id=booking.booking_id,
                    seat_code=seat_code,
                    seat_price=available_seats[seat_code]["price"],
                )
            )

        db.session.commit()
        _release_seats(showtime_id, unique_seat_codes, access_token)
        return booking
    except Exception:
        db.session.rollback()
        # Release the seat locks if booking fails
        _release_seats(showtime_id, unique_seat_codes, access_token)
        raise


def get_booking_history(user_id):
    return (
        Booking.query.filter_by(user_id=user_id)
        .order_by(Booking.booking_id.desc())
        .all()
    )


def get_reserved_seats(showtime_id):
    reserved_rows = (
        db.session.query(BookingDetail.seat_code)
        .join(Booking, Booking.booking_id == BookingDetail.booking_id)
        .filter(
            Booking.showtime_id == showtime_id,
            Booking.status >= 0,
        )
        .distinct()
        .order_by(BookingDetail.seat_code.asc())
        .all()
    )

    return [row[0] for row in reserved_rows]


def get_booking_detail(booking_id, user_id):
    booking = Booking.query.filter_by(booking_id=booking_id, user_id=user_id).first()
    if not booking:
        return None, []

    details = BookingDetail.query.filter_by(booking_id=booking_id).all()
    return booking, details

File: microservices C#/order_service/Services/BookingService.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using OrderService.Models;

namespace OrderService.Services;

public sealed class BookingService(OrderDbContext db, IHttpClientFactory httpClientFactory)
{
    public async Task<Booking> CreateBooking(int userId, BookingRequest request, string accessToken)
    {
        var showtimeId = request.ShowtimeId;
        var seatCodes = ExtractSeatCodes(request).Distinct().ToList();
        if (showtimeId <= 0 || seatCodes.Count == 0)
            throw new BookingException("showtime_id and seat_codes are required");

        var cinema = CreateCinemaClient(accessToken);
        var lockResponse = await cinema.PostAsJsonAsync($"showtimes/{showtimeId}/seats/lock", new { seat_codes = seatCodes });
        if (!lockResponse.IsSuccessStatusCode)
            throw new BookingException($"Unable to lock seats: {await UpstreamMessage(lockResponse)}");

        try
        {
            var availableSeats = await FetchAvailableSeats(cinema, showtimeId);
            var invalidSeats = seatCodes.Where(code => !availableSeats.ContainsKey(code)).ToList();
            if (invalidSeats.Count > 0)
                throw new BookingException($"Invalid seat codes: {string.Join(", ", invalidSeats)}");

            var alreadyBooked = await GetReservedSeats(showtimeId, seatCodes);
            if (alreadyBooked.Count > 0)
                throw new BookingException($"Seats already booked: {string.Join(", ", alreadyBooked.OrderBy(code => code))}");

            var booking = new Booking
            {
                UserId = userId,
                ShowtimeId = showtimeId,
                TotalAmount = seatCodes.Sum(code => availableSeats[code]),
                Status = 1,
                CreatedAt = DateTime.UtcNow
            };
            db.Bookings.Add(booking);
            await db.SaveChangesAsync();

            foreach (var code in seatCodes)
            {
                db.BookingDetails.Add(new BookingDetail
                {
                    BookingId = booking.BookingId,
                    SeatCode = code,
                    SeatPrice = availableSeats[code]
                });
            }

            await db.SaveChangesAsync();
            return booking;
        }
        finally
        {
            await cinema.PostAsJsonAsync($"showtimes/{showtimeId}/seats/release", new { seat_codes = seatCodes });
        }
    }

    public Task<List<Booking>> GetHistory(int userId) =>
        db.Bookings
            .Where(booking => booking.UserId == userId)
            .OrderByDescending(booking => booking.BookingId)
            .ToListAsync();

    public Task<List<string>> GetReservedSeats(int showtimeId) =>
        db.BookingDetails
            .Join(db.Bookings, detail => detail.BookingId, booking => booking.BookingId, (detail, booking) => new { detail, booking })
            .Where(row => row.booking.ShowtimeId == showtimeId && row.booking.Status >= 0)
            .Select(row => row.detail.SeatCode)
            .Distinct()
            .OrderBy(code => code)
            .ToListAsync();

    public async Task<(Booking? Booking, List<BookingDetail> Details)> GetBookingDetail(int bookingId, int userId)
    {
        var booking = await db.Bookings.FirstOrDefaultAsync(item => item.BookingId == bookingId && item.UserId == userId);
        if (booking is null)
            return (null, []);

        var details = await db.BookingDetails.Where(detail => detail.BookingId == bookingId).ToListAsync();
        return (booking, details);
    }

    private HttpClient CreateCinemaClient(string accessToken)
    {
        var cinema = httpClientFactory.CreateClient("cinema");
        cinema.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
        return cinema;
    }

    private async Task<Dictionary<string, decimal>> FetchAvailableSeats(HttpClient cinema, int showtimeId)
    {
        var seatMapResponse = await cinema.GetAsync($"showtimes/{showtimeId}/seats");
        if (seatMapResponse.StatusCode == System.Net.HttpStatusCode.NotFound)
            throw new BookingException("Showtime not found");
        if (!seatMapResponse.IsSuccessStatusCode)
            throw new BookingException("Unable to verify seats from cinema service");

        var seatMapJson = await seatMapResponse.Content.ReadFromJsonAsync<JsonElement>();
        var seatsRoot = seatMapJson.TryGetProperty("seats", out var directSeats)
            ? directSeats
            : seatMapJson.GetProperty("data").GetProperty("seats");
        return seatsRoot.EnumerateArray()
            .ToDictionary(item => item.GetProperty("code").GetString() ?? "", item => item.GetProperty("price").GetDecimal());
    }

    private Task<List<string>> GetReservedSeats(int showtimeId, IReadOnlyCollection<string> seatCodes) =>
        db.BookingDetails
            .Join(db.Bookings, detail => detail.BookingId, booking => booking.BookingId, (detail, booking) => new { detail, booking })
            .Where(row => row.booking.ShowtimeId == showtimeId && row.booking.Status >= 0 && seatCodes.Contains(row.detail.SeatCode))
            .Select(row => row.detail.SeatCode)
            .ToListAsync();

    private static List<string> ExtractSeatCodes(BookingRequest request)
    {
        if (request.SeatCodes is { Count: > 0 })
            return request.SeatCodes.Where(code => !string.IsNullOrWhiteSpace(code)).Select(code => code.Trim().ToUpperInvariant()).ToList();
        if (request.Seats is null)
            return [];

        var result = new List<string>();
        foreach (var seat in request.Seats)
        {
            if (seat.ValueKind == JsonValueKind.String && !string.IsNullOrWhiteSpace(seat.GetString()))
                result.Add(seat.GetString()!.Trim().ToUpperInvariant());
            if (seat.ValueKind == JsonValueKind.Object && seat.TryGetProperty("seat_code", out var seatCode) && !string.IsNullOrWhiteSpace(seatCode.GetString()))
                result.Add(seatCode.GetString()!.Trim().ToUpperInvariant());
        }
        return result;
    }

    private static async Task<string> UpstreamMessage(HttpResponseMessage response)
    {
        try
        {
            var json = await response.Content.ReadFromJsonAsync<JsonElement>();
            return json.TryGetProperty("message", out var message) ? message.GetString() ?? "Unable to lock seats" : "Unable to lock seats";
        }
        catch
        {
            return "Unable to lock seats";
        }
    }
}

